### PR TITLE
Use parameter binding in TDEngine target insertions

### PR DESCRIPTION
**Pull request by jond01**
See:
https://docs.tdengine.com/develop/insert-data/sql-writing/#insert-with-parameter-binding
https://docs.tdengine.com/reference/connectors/python/#writing-data-via-parameter-binding

Fixes [ML-7662](https://iguazio.atlassian.net/browse/ML-7662).

Improve performance and security, e.g. protection against SQL injections, with parameter-bound `INSERT` transaction.

[ML-7662]: https://iguazio.atlassian.net/browse/ML-7662?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: storey/targets.py [hunk 1]
```diff
@@ -888,48 +891,100 @@ def _init(self):
...
+            for val_ind, val_name in enumerate(val_names):
+                raw_value = event.get(val_name)
+                if event_ind == 0:
+                    to_column_funs.append(self._raw_value_to_column_fun(raw_value))
+                values[val_ind].append(self._raw_value_to_value(raw_value))
+        return [fun(vals) for fun, vals in zip(to_column_funs, values)]
+
+    async def _emit(self, batch: list[dict], batch_key: str, batch_time, batch_events, last_event_time=None):
+        stmt = self._connection.statement()
+        stmt.prepare(self._sql_template)
```
Comment (gtopper) [inline storey/targets.py:1022]: Please move prepared statement to self for better performance.

File: integration/test_tdengine.py [hunk 2]
```diff
@@ -133,3 +134,49 @@ def test_tdengine_target(tdengine, table_col):
...
+        [
+            SyncEmitSource(),
+            TDEngineTarget(
+                url=url,
+                time_col="time",
+                columns=["my_string"],
+                user=user,
+                password=password,
+                database=db_name,
+                table=None,
```
Comment (gtopper) [inline integration/test_tdengine.py]:
```suggestion
```

File: storey/targets.py [hunk 3]
```diff
@@ -805,6 +805,9 @@ class TDEngineTarget(_Batching, _Writer):
     :type flush_after_seconds: int
     """
 
+    class UnsupportedTDEngineTypeError(TypeError):
+        pass
```
Comment (gtopper) [inline storey/targets.py]: I don't think we need a custom error for that. If we do, a better place for it is probably in `dtypes.py` with the other custom errors. There's no need for it to be nested (which is something I don't think we do anywhere in storey or even in mlrun).

File: storey/targets.py [hunk 4]
```diff
@@ -888,48 +891,100 @@ def _init(self):
...
+        raise cls._get_unsupported_error(value)
+
+    def _get_tags_from_event(self, event: dict) -> list:
+        tags = []
+        for tag_name in self._columns[: self._number_of_tags]:
+            tags.append(self._value_to_tag(event.get(tag_name)))
+        return tags
+
+    @classmethod
+    def _raw_value_to_column_fun(cls, value) -> Callable:
```
Comment (gtopper) [inline storey/targets.py]:
```suggestion
    def _raw_value_to_column_func(cls, value) -> Callable:
```
(or function)

Applies elsewhere as well.

File: storey/targets.py [hunk 5]
```diff
@@ -888,48 +891,100 @@ def _init(self):
...
+            return int(value.timestamp() * 1000)
+        return value
+
+    def _get_batch_values(self, batch: list[dict]) -> list:
+        values = [[] for _ in range(self._number_of_values)]
+        to_column_funs = []
+        val_names = self._columns[self._number_of_tags :]
+        for event_ind, event in enumerate(batch):
+            for val_ind, val_name in enumerate(val_names):
+                raw_value = event.get(val_name)
```
Comment (gtopper) [inline storey/targets.py]: Need to raise an error if missing (or is that ok in some cases – i.e. nullable columns?). As it's written, it will raise a
```UnsupportedTDEngineTypeError("Unsupported value type NoneType in TDEngineTarget")```

File: storey/targets.py [hunk 6]
```diff
@@ -888,48 +891,100 @@ def _init(self):
...
+    @classmethod
+    def _raw_value_to_value(cls, value):
+        if isinstance(value, datetime.datetime):
+            return int(value.timestamp() * 1000)
+        return value
+
+    def _get_batch_values(self, batch: list[dict]) -> list:
+        values = [[] for _ in range(self._number_of_values)]
+        to_column_funs = []
+        val_names = self._columns[self._number_of_tags :]
```
Comment (gtopper) [inline storey/targets.py]: I don't think "value names" is correct nomenclature.
```suggestion
        col_names = self._columns[self._number_of_tags :]
```
or `column_names`. Or `regular_column_names` I suppose.

File: storey/targets.py [hunk 7]
```diff
@@ -888,48 +891,100 @@ def _init(self):
...
+        for event_ind, event in enumerate(batch):
+            for val_ind, val_name in enumerate(val_names):
+                raw_value = event.get(val_name)
+                if event_ind == 0:
+                    to_column_funs.append(self._raw_value_to_column_fun(raw_value))
+                values[val_ind].append(self._raw_value_to_value(raw_value))
+        return [fun(vals) for fun, vals in zip(to_column_funs, values)]
```
Comment (gtopper) [inline storey/targets.py]: Seems unnecessarily functional to me. I'd do something like
```python
    def _get_batch_values(self, batch: list[dict]) -> list:
        result = []
        val_names = self._columns[self._number_of_tags :]
        for event in batch:
            values = []
            for val_name in val_names:
                raw_value = event.get(val_name)
                values.append(self._raw_value_to_value(raw_value))
            to_column_func = self._raw_value_to_column_fun(event.get(val_names[0]))
            result.append(to_column_func(values))
        return result
```
(potentially dealing with `event.get(val_names[0])` differently, and do note that I didn't test this snippet)

File: storey/targets.py [hunk 8]
```diff
@@ -888,48 +891,100 @@ def _init(self):
...
+        if isinstance(value, float):
+            return taosws.float_to_tag(value)
+        if isinstance(value, str):
+            return taosws.varchar_to_tag(value)
+        raise cls._get_unsupported_error(value)
+
+    def _get_tags_from_event(self, event: dict) -> list:
+        tags = []
+        for tag_name in self._columns[: self._number_of_tags]:
+            tags.append(self._value_to_tag(event.get(tag_name)))
```
Comment (gtopper) [inline storey/targets.py]: Here too, we need to handle a missing value better than raising an `Unsupported value type NoneType` error.

**Review by gtopper — COMMENTED** [review]
Nice. Some comments but looks good overall.

For reference, this is the PR that added parameter binding in mlrun – https://github.com/mlrun/mlrun/pull/6195.

File: storey/targets.py (hunk 1 above)
Comment (jond01) [reply storey/targets.py:1022]: I tried reusing a statement, but it cannot "clean" previous bindings - it only aggregates the new parameters.
So I think we should instantiate a new statement for each emission.

File: storey/targets.py (hunk 5 above)
Comment (jond01) [reply storey/targets.py]: I'll handle the type resolution with the table's schema - obtained through a `DESCRIBE` query.

File: storey/targets.py (hunk 1 above)
Comment (gtopper) [reply storey/targets.py:1022]: Ok, maybe there's a way to get it working, but we can put it aside for now.

File: storey/targets.py (hunk 5 above)
Comment (jond01) [reply storey/targets.py]: Done

File: storey/targets.py [hunk 9]
```diff
@@ -888,48 +922,87 @@ def _init(self):
...
+        self, table_name: str
+    ) -> tuple[
+        list[tuple[str, Callable[[Any], "taosws.PyTagView"]]], list[tuple[str, Callable[[list], "taosws.PyColumnView"]]]
+    ]:
+        fields_data = [_TDEngineFieldData(*raw) for raw in self._connection.query(f"DESCRIBE {table_name};")]
+        tags_schema = []
+        reg_cols_schema = []
+        for field_data in fields_data:
+            name = field_data.field
+            type_ = field_data.type
```
Comment (gtopper) [inline storey/targets.py]: Maybe call it `field_type`.

File: storey/targets.py [hunk 10]
```diff
@@ -888,48 +922,87 @@ def _init(self):
...
+            try:
+                if field_data.note == "TAG":
+                    tags_schema.append((name, self._tdengine_type_to_tag_func[type_]))
+                else:
+                    reg_cols_schema.append((name, self._tdengine_type_to_column_func[type_]))
+            except KeyError:
```
Comment (gtopper) [inline storey/targets.py]: It's better check using a condition rather than causing a `KeyError` and catching it, from a performance standpoint.

File: storey/targets.py [hunk 11]
```diff
@@ -888,48 +922,87 @@ def _init(self):
...
+                raise TDEngineTypeError(
+                    f"Unsupported value type '{type_}' of field '{name}' in {self.__class__.__name__}"
```
Comment (gtopper) [inline storey/targets.py]:
```suggestion
                    f"Unsupported value type '{type_}' of field '{name}' in {type(self).__name__}"
```

File: storey/targets.py [hunk 12]
```diff
@@ -888,48 +922,87 @@ def _init(self):
...
-        return str(value)
+    def _get_params_template(num_param: int) -> str:
+        return f"({','.join(num_param * ['?'])})"
+
+    def _get_table_schema(
+        self, table_name: str
+    ) -> tuple[
+        list[tuple[str, Callable[[Any], "taosws.PyTagView"]]], list[tuple[str, Callable[[list], "taosws.PyColumnView"]]]
+    ]:
+        fields_data = [_TDEngineFieldData(*raw) for raw in self._connection.query(f"DESCRIBE {table_name};")]
```
Comment (gtopper) [inline storey/targets.py]: I guess we should validate the table name in the constructor then?

File: storey/targets.py [hunk 13]
```diff
@@ -888,48 +922,87 @@ def _init(self):
...
+        tags_schema = []
+        reg_cols_schema = []
+        for field_data in fields_data:
```
Comment (gtopper) [inline storey/targets.py]: Maybe just call it
```suggestion
        for field in fields:
```
because `fields_data` isn't correct grammar, which makes it harder to understand.

File: storey/targets.py [hunk 14]
```diff
@@ -888,48 +922,87 @@ def _init(self):
...
+    @staticmethod
+    def _get_tags_from_event(
+        tags_schema: list[tuple[str, Callable[[Any], "taosws.PyTagView"]]], event: dict
+    ) -> list["taosws.PyTagView"]:
+        return [tag_func(event.get(tag_name)) for tag_name, tag_func in tags_schema]
+
+    @staticmethod
+    def _raw_value_to_value(value):
+        if isinstance(value, datetime.datetime):
+            return int(value.timestamp() * 1000)  # use millisecond precision
```
Comment (gtopper) [inline storey/targets.py]: I'd clarify to avoid implying that we have any other choice.
```suggestion
            return int(value.timestamp() * 1000)  # convert to milliseconds
```

File: tests/test_types.py [hunk 15]
```diff
@@ -76,3 +77,21 @@ def test_emit_policy_period():
     policy = _dict_to_emit_policy(policy_dict)
     assert type(policy) == EmitAfterPeriod
     assert policy.delay_in_seconds == 8
+
+
+@pytest.mark.parametrize(
+    "description",
+    [
+        # taosBenchmark
+        # list(conn.query("describe test.d6241;"))
```
Comment (gtopper) [inline tests/test_types.py]: What does it mean?

File: tests/test_types.py [hunk 16]
```diff
@@ -76,3 +77,21 @@ def test_emit_policy_period():
...
+        ("ts", "TIMESTAMP", 8, "", "delta-i", "lz4", "medium"),
+        ("current", "FLOAT", 4, "", "delta-d", "lz4", "medium"),
+        ("voltage", "INT", 4, "", "simple8b", "lz4", "medium"),
+        ("phase", "FLOAT", 4, "", "delta-d", "lz4", "medium"),
+        ("groupid", "INT", 4, "TAG", "disabled", "disabled", "disabled"),
+        ("location", "VARCHAR", 24, "TAG", "disabled", "disabled", "disabled"),
+    ],
+)
+def test_description_parsing_to_tdengine_field(description: tuple) -> None:
+    """Test that the parsing works"""
```
Comment (gtopper) [inline tests/test_types.py]: Not sure I understand what parsing this is testing. If it does test parsing, where are the error cases?

File: storey/targets.py [hunk 17]
```diff
@@ -888,48 +922,87 @@ def _init(self):
...
+    def _get_batch_values(
+        cls, reg_cols_schema: list[tuple[str, Callable[[list], "taosws.PyColumnView"]]], batch: list[dict]
+    ) -> list["taosws.PyColumnView"]:
+        return [
+            col_func([cls._raw_value_to_value(event.get(col_name)) for event in batch])
+            for col_name, col_func in reg_cols_schema
+        ]
+
+    async def _emit(self, batch: list[dict], batch_key: str, batch_time, batch_events, last_event_time=None):
+        tags_schema, reg_cols_schema = self._get_table_schema(self._table or self._supertable)
```
Comment (gtopper) [inline storey/targets.py]: We shouldn't have to describe the table before each statement, doubling the number of IO operations.

File: storey/targets.py [hunk 18]
```diff
@@ -888,48 +922,87 @@ def _init(self):
...
+        if isinstance(value, datetime.datetime):
+            return int(value.timestamp() * 1000)  # use millisecond precision
+        return value
+
+    @classmethod
+    def _get_batch_values(
+        cls, reg_cols_schema: list[tuple[str, Callable[[list], "taosws.PyColumnView"]]], batch: list[dict]
+    ) -> list["taosws.PyColumnView"]:
+        return [
+            col_func([cls._raw_value_to_value(event.get(col_name)) for event in batch])
```
Comment (gtopper) [inline storey/targets.py:1016]: It's a little hard to follow, but what happens if `event.get(col_name)` resolves to `None`?

File: storey/targets.py (hunk 14 above)
Comment (jond01) [reply storey/targets.py]: But we do - we could use micro-sec or nano-sec:
https://docs.tdengine.com/reference/taos-sql/data-type/#timestamp

https://docs.tdengine.com/cloud/taos-sql/data-type/#data-types:
> Default precision is millisecond, microsecond and nanosecond are also supported.

File: tests/test_types.py (hunk 15 above)
Comment (jond01) [reply tests/test_types.py]: It comes from the benchmark DB - specifically this table. I think it can be useful as a reference

File: storey/targets.py (hunk 18 above)
Comment (jond01) [reply storey/targets.py:1016]: Then it is converted via `col_func`, and inserted to the table as `NULL`.

File: tests/test_types.py (hunk 16 above)
Comment (jond01) [reply tests/test_types.py]: Only happy flows are tested here.
But for example - more/fewer fields than expected will result in an error.

File: storey/targets.py (hunk 14 above)
Comment (gtopper) [reply storey/targets.py]: I meant we don't have a choice at this point in the code :)

File: tests/test_types.py (hunk 16 above)
Comment (gtopper) [reply tests/test_types.py]: That can be said for any constructor or function that doesn't take `*args` though. It doesn't really count as parsing. And doesn't really justify a test if I'm not misunderstanding something here.

File: storey/targets.py (hunk 14 above)
Comment (jond01) [reply storey/targets.py]: This part does not create the table so it does not decide on precision.
However, the table's creator has the choice, but we only support ms at the moment.

File: tests/test_types.py (hunk 16 above)
Comment (jond01) [reply tests/test_types.py]: This test is beneficial for the following reasons:
1. Increases the coverage in the non-integration tests that run in the CI. TDEngine integration tests are skipped.
2. Insures a regression in the class - e.g. removing/adding a field - will be caught.

File: tests/test_types.py (hunk 15 above)
Comment (jond01) [reply tests/test_types.py]: https://docs.tdengine.com/get-started/docker/#test-data-insert-performance

File: tests/test_types.py (hunk 16 above)
Comment (gtopper) [reply tests/test_types.py]: Would be nice to turn it into a test of actual functionality by testing the logic in `_get_table_schema` (preferably by moving everything other than `self._connection.query(...)` to another method and testing that).

Then it can test error cases as well.

**Review by gtopper — APPROVED** [review]
